### PR TITLE
Feature: registered parameters

### PR DESCRIPTION
**Pull request by dertseha**
This PR is the result of #37 .

Thinking further, I figured that the registered parameters could be tracked by Go's built-in flags system. This handles usage output, as well as invalid parameter handling out of the box; As well as having less code to handle within this library.

Example output of `go run build\build.go -h` (output shortened to highlight diff):
```
Usage: [flag(s)] task(s)
Flags:
  -ci string
        Whether CI is calling the build script (default "false")
  -v    Verbose output: log all tasks as they are run. Also print all text from Log and Logf calls even if the task succeeds.
Tasks:
  ...
  diff        git diff; -ci
  ...
```

This is a breaking change as Taskflow.Params is dropped and tasks *need* to specify which parameters they read (a change as intended).

More thoughts on these parameters:
* I wish the type information could be saved. As it is now, implementation of `TFParams` mirrors all the parsing what `flag` package has internally for the typed parameters. As it stands, the current setup does not support keeping the type information. Unless the type information of the parameter (string, int, ...) is carried over to a `RegisteredParamString`, `RegisteredParamInt`, and so forth - though this generates even more code.
* Is there big of a need of a built-in support for text/json parsing of parameters? This sounds a bit overengineered to me.

Most of the tests could carry over or needed only slight modification. `Configure` and `MustConfigure` aren't covered yet.

How does this look?

### CODE REVIEW

**Comment by dertseha** [comment]
(waiting on your thoughts before chasing 100% coverage)

**Comment by pellared** [comment]
First of all thanks a lot for this PR and sorry for the delay. I wanted to experiment with it before giving feedback too fast.

## Reusing flags

I feel reusing flags here would not be a good decision. Not from an implementation perspective. But from a usage perspective.

1. I would rather keep flags for stuff like `-h` for help, `-v` for verbose, `-graph` to print a dependency graph
1. `goyek -v test test-mode=short lint lint-mode=fast` would be more readable (at least for me)
1. Having help output structured more or less like below would be more readable:

```
Usage: [flag(s)] task(s) [param=value]
Default task: all
Flags:
  -v    Verbose output: log all tasks as they are run. Also, print all text from Log and Logf calls even if the task succeeds.
Tasks:
  all       build pipeline
  build       go build
  clean       remove git ignored files
  diff (ci)       git diff
  fmt         gofumports
  install     install build tools
  lint        golangci-lint
  mod-tidy    go mod tidy
  test        go test with race detector and code coverage
Default task: all
Params:
  ci [bool, default: false]    Whether CI is calling the build script
```

## Params registration and usage

It would be good to couple the API related with parameter registration with usage. It would be cool to achieve something like:

```
func taskDiff(ci taskflow.ParamBool) taskflow.Task {
	return taskflow.Task{
		Name:        "diff",
		Parameters:  taskflow.Params{ ci },
		Command: func(tf *taskflow.TF) {
			if !(ci.Bool(tf)) {
				tf.Skip("ci param is not set, skipping")
			}
			// ....
		},
```

`ci.Bool(tf)` should call `tf.Fatalf` if it would run in a task where the parameter was not registered.


##  Text/json parsing of parameters

> Is there big of a need of a built-in support for text/json parsing of parameters? This sounds a bit overengineered to me.

No. However, it would be good to do something like https://golang.org/pkg/flag/#Var + https://golang.org/pkg/flag/#Getter to make the package flexible. Can be always added as separate PR.

**Comment by dertseha** [comment]
Thank you for your response, a good basis for a discussion.

## Params registration and usage

To clarify, in your code example, I suspect you meant to write

```
...
		Command: func(tf *taskflow.TF) {
			if !ci.Bool(tf) {
...
```
(passing `tf` to the `ci.Bool()` as well) -- otherwise the created parameter needs to capture access to the - at its point of creation future instance of - `tf` before, and you have two parallel usages of `tf`, one explicit and one implicit.
Though, in your examples you explicitly state the capture.

> I guess we would need to have types like `type Param struct { tf * *taskflow.TF, ... } `, `type Params []Param`, `type ParamBool { Param, ... }` to achieve such API.

This is what I meant with more code necessary, yes. 

## Reusing flags

To me it looks like you want to separate the things you want from core `goyek` from the freely defined tasks.
I however ask why making extra effort to separate that when it can be based on the same framework?

From your samples I first want to differentiate between:
* providing of help, and later generating a dependency graph
* having a "common" option for "verbose" output.

I pose the question whether execution of `help` or `graph` be any different from another task like `ci` or `test`? I could consider them as tasks like any other.

For example, what if core `goyek` would only know how to execute tasks with parameters, and provide an API for introspection. Then the standard package provides standard tasks, called `help` and later `graph`. They could be available as either an opt-in or opt-out model (API to do that to be determined). These tasks would then be implemented by using the introspection API.

Or, said differently: your proposed `-graph` *does* something (a task, in my view) and `-v` *modifies* executions - two different things via the same interface style.


As for the common request of "verbose" output, I see this equal to any other parameter for the tasks. If I want verbose output of the `lint` task, why do I have to specify it via a flag, if I request the linter mode via an argument? These are two different ways to modify the execution of one task.


Please also refer to a cool resource I recently read: [Command Line Interface Guidelines](https://clig.dev/); To the current point I'd like to point out the following two sections:

First, what brought me to the idea of offering help as a task as well, not just as a flag:

> If your program is git-like, the following should also offer help:
> `myapp help`

Found in section about [Help](https://clig.dev/#help).

And second, "Prefer flags to args." found in [Arguments and flags](https://clig.dev/#arguments-and-flags)

(This guide could in general give good points on what should be considered in `goyek`, for instance about output)

## Text/JSON parsing of parameters

With the additional types for parameters this should be extensible. I believe - assuming the route with flags is continued - `goeyk` would provide wrappers to the underlying flag types and thus also offer a `Value` based approach.

**Comment by pellared** [comment]
## Params registration and usage

Just to clarify, I was thinking of capturing the `*taskflow.TF` instance somewhere around this place: https://github.com/pellared/taskflow/blob/ddea2cc79a5db5e700c7eb796818812a6f9a2865/flow_runner.go#L161 so that the usage of `tf` could be implicit.

## Reusing flags

_First of all, I  need to do my homework and read https://clig.dev/ . All that is written below are my first impressions and initial feedback. I can change my mind after some time when I start understanding your concerns better._

### help/graph as tasks

> I pose the question whether execution of `help` or `graph` be any different from another task like `ci` or `test`? I could consider them as tasks like any other.

I separated it b/c I had problems with printing to output. The best I could do is to write help to `os.Stdout` instead of `tf.Output` to get a result like: 

```
[HELP OUTPUT]
ok      0.021s
```

I did not like the `ok      0.021s` ending. But I am getting convinced that having `help` as the task is more logical (even with this help `OK` ending. I assume that we should then create some common `help` task that
t could be used by the user. When we add API for introspection then anyone can easily create its own `help` tasks (e.g. with colours).

> [...] provide an API for introspection

I was thinking of it even before. I just did not want to expose too much API upfront and I felt it is something that can be easily added at any point. This kind of API could e.g. also facilitate making some wrappers for taskflow/goyek. I imagine it could be even possible to create a little web application based on tasks, parameters etc. I think this is something that should grow in a natural way. E.g. when working on making `help` as a task.

**Summing up**, I feel convinced that `help` and `graph` should be defined and working as tasks. They should not be "flags". @dertseha Are we on the same page?


### verbose output

> As for the common request of "verbose" output, I see this equal to any other parameter for the tasks. If I want verbose output of the lint task, why do I have to specify it via a flag, if I request the linter mode via an argument? These are two different ways to modify the execution of one task.

I feel that I do not understand this fragment. I will present my understanding of `-v` flag and why it is not a regular parameter.

I see `-v` flag is more like the taskflow output moderather than a task parameter. E.g. I would NOT suggest to use `tf.Verbose()` to determine that a linter should be run with verbose output. If I would want to run a linter with verbose output then it could be something like: `goyek lint lint-args="-v"` and to run goyek as verbose and linter as verbose as well: `goyek -v lint lint-args="-v"`. I just feel it is good to differentiate in CLI syntax: the taskflow configuration (flags with `-` prefix), tasks (just names) and task parameters (key`=`val). 

@dertseha, please let me know if it makes any sense to you?

## Similar libraries in .NET ecosystem

About 2 months ago I asked colleagues if there is something similar to taskflow in .NET ecosystem. And of course, it occurred that something similar already exists. 

- https://github.com/adamralph/bullseye#command-line-arguments
- https://nuke.build/docs/running-builds/global-tool.html 

I started to look at its design for inspiration.

**Comment by dertseha** [comment]
Thank you again; I'll wait further until you have gathered other information. For now responses to clarify points. And thank you for being open to other opinions :)

> ## task parameters design
> Just to clarify, I was thinking of capturing the `*taskflow.TF` instance somewhere around this place:
>  ...
> so that the usage of `tf` could be implicit.

My concern is less about the implementation - that I can do; I was more concerned about the design choice: The task of the command receives `tf` as an argument, which the task is only allowed to use during its execution. At the same time, the task also has references to the registered values, which implicitly access (the same) `tf`.
As these registered values are present already before the whole task tree is run, a crafty user could attempt to call `param.Bool()` outside any `tf` context. Of course it would need to crash/report error/... , yet it seems odd to me that the user has not only the restriction of using `tf` only within that task function, but this restriction also for the registered values.

I understand the technical reason that this is the case - however the API does not make this explicit. This, compared to: "Ah, to query the parameter, I need a `tf` reference - which I only have within that function. So I can query it only there."

> ## Reusing flags
> ### verbose output
> I feel that I do not understand this fragment. I will present my understanding of `-v` flag and why it is not a regular task parameter.

I'm with you that there should not be a `tf.Verbose()` call within the task. I would even go as far as to remove that function.
I believe we have a mixup of our understanding of terminology. I'll back up a few steps and describe how I see the usage, including your examples and extending it.

From the outside - as a user on the CLI - I do not consider `goyek` itself. I only think in terms of tasks. Furthermore, any flag that the current set of registered tasks use, I consider "global". Maybe that some flags are very specific to single tasks, typically expressed in their name, and other flags affect several or all tasks.

For the following examples, I'll take the topic of verbosity again, with one available task `lint`.

`goyek lint` would run the task and print all the details, including the introduction that `lint` is being run, and that it completed, possibly including timing (see other thread, below).

If I wanted a verbose output overall, so I receive only the most pressing linter errors and nothing else, I could set up the user Go code that allows me to run `goyek lint -v`, which does not print the introduction/summary, as well as only the most immediate linter issues. In the background (the user Go code), the `v` parameter was registered and given to both goyek and the lint task to specify verbose output.

If I wanted to support a separate configuration that allows me to print all the linter reports, but not the task introduction/summary, I could do something in the user Go code to support `goyek lint -raw`. "Raw" is again a registered parameter in the user Go code and given only to goyek, like in the example above. "Raw" would mean: give me only the output from the task(s), nothing more.

To add information for both previous cases: `goyek` API would have something like `flow.SetVerbosityParam(registeredParam)` and not care whether it is called `v`, `raw`, or `please-do-not-log-task-info`).

If I wanted to also support a user Go code provided verbosity flag exclusively for the linter, I could add something to support `goyek lint -linter-mode verbose`. In the background (user Go code) this then maps the value of `verbose` of the registered parameter `linter-mode` to the corresponding flag of the actual linter process execution (`-v`).

If I wanted to support arbitrary parameters to be forwarded to the linter process, this could be done with registered parameter `linter-args`, and used like this: `goyek lint -linter-args "--all --long ..."`. This registered parameter is given to the linter process verbatim.

This allows the same API for all cases to the CLI-user (it's always only flags); The Go code dev (the one wiring up the tasks) has full control whether to re-use registered parameters in various places - or not; And the `goyek` code itself becomes more streamlined.

Of course some parameter, most notably the one about verbosity, are common enough that the could be provided as standard registered parameter, available in a similar way to the standard `help` task. Still, the Go code dev has the option to use them (API to be determined) or come up with their own.

Finally from a different angle: With the agreement of having `help` and `graph` being tasks, there's only the one `-v` left in the current design. This means the familiarity (and existing code, Go with batteries included) of flags is used up for one parameter, and the majority of goyek uses for parameters of the tasks is a separate, uncommon system, that has to be implemented.
My thinking here is to better use common code/patterns for the common use-cases (task parameters done with flags), and think of special solutions only for edge cases (make a way to have `goyek` library take a registered parameter as a control about its verbosity).

I was not sure whether or not I should add code sample as well and whether they would make things clearer how I mean them - or they would add more noise. Please tell if I should add some.

> ### help/graph as tasks
> **Summing up**, I feel convinced that `help` and `graph` should be defined and working as tasks. They should not be "flags". @dertseha Are we on the same page?

Yes, and I believe even the output of timing information could be under discussion. With the existence of other shell commands like [time](https://linux.die.net/man/1/time), perhaps the measurement of execution time does not have to be implemented in `goyek`.
(I know this is out of scope for the PR at hand as well, it's meant as a further option to think about.)

In summary, in all the points above I am critically thinking along the way of interface design: Perfection is achieved not if you can't find anything more to add, but when you can't find anything more to remove.

**Comment by pellared** [comment]
## task parameters design

> My concern is less about the implementation - that I can do; I was more concerned about the design choice: The task of the command receives `tf` as an argument, which the task is only allowed to use during its execution. At the same time, the task also has references to the registered values, which implicitly access (the same) `tf`.
> As these registered values are present already before the whole task tree is run, a crafty user could attempt to call `param.Bool()` outside any `tf` context. Of course it would need to crash/report error/... , yet it seems odd to me that the user has not only the restriction of using `tf` only within that task function, but this restriction also for the registered values.
> 
> I understand the technical reason that this is the case - however the API does not make this explicit. This, compared to: "Ah, to query the parameter, I need a `tf` reference - which I only have within that function. So I can query it only there."

I agree with you. 👍 

## Reusing flags

I feel your ideas are worth giving a try 😉 

Only a few remarks which I am not sure if were mentioned:
1. It would be good to specify the parameter after the task e.g. `goyek -v all -ci`. It is not working in current PR.
2. In `help` it would be good to separate OOTB taskflow parameters (currently only `-v`, however, I already feel a need to add `dir` which will [change the working directory](https://golang.org/pkg/os/#Chdir)) from tasks' parameters

> I was not sure whether or not I should add code sample as well and whether they would make things clearer how I mean them - or they would add more noise. Please tell if I should add some.

Your explanations are clear (probably a lot more than mine). 

I propose to update the current PR with some of our ideas. It can still be a PoC PR without coverage and with only support for `bool` and `string` parameters. I suggest first tackling the "task parameters". 

The other ideas including (but not limited to):
1. redesign "taskflow parameters" (currently verbose)
2. reimplementing `help` as a task

Can be done later. 

But if you prefer to make some PoC with all your proposals I am also OK with it. Whatever works for you.

**Comment by dertseha** [comment]
A work in progress comment.

With the newest change, I reworked the flags to have also "verbose" be a regular parameter, as well as having typed parameters that would also allow a JSON extension.
This implementation now strongly relies on the "native" implementation of the `flag` package.

This, sadly, also implies that all flags have to be provided before the task names, as the standard package aborts flag parsing as soon as a non-flag is given.
I understand that this is a needed feature in order to allow passing task-specific flags next to their name, possibly between other task names. Also, CLIG as well notes that flags should be allowed at any position.
I'll now have another pass that I somewhat wanted to avoid: reintroduce flag parsing as the included batteries don't fit :/
At least this would allow using double-dashes as well.

**Comment by dertseha** [comment]
Now without standard `flag` package - tasks and flags can occur in any sequence. Additional benefit: double dash (`--`) and short form (`-`) could be implemented - see for `verbose`.

Usage prints flags unsorted. I figure there might be extra demands, such as show task-exclusive parameters only in their section - and have a "common" section only for those that multiple tasks use.

As for usage itself, when special cases like `-h` and `--help` also should be supported, then this would require an extra workaround to support task-via-flag start...

**Comment by dertseha** [comment]
I've now also included test coverage. For me this looks like a good basis for further extensions, which are all around introspection: what tasks/parameters are there, is every registered parameter used, and so on. The entire introspection part could be a separate PR.

As the code is now, task `help` *could* be overridden by user code, as well as `-h` and `--help`.
As per CLIG, the two flags *should not* be redefined, and should always be implemented. So I'm OK with the current solution that these two flags would always provide usage information. Based on the future introspection API, a dedicated help task could be made, which internally refers to the same code.

Another thing I did implicitly: I removed the separation betwee `Configure*` and `MustConfigure*`. Reasoning: For now a panic suffices, and *if* error handling is required, this could be moved as a general setup-check for `Run()`, and configured in taskflow.New(). Perhaps this is also something to consider for task registration.
Error-handling on a function level would duplicate all these type-specific functions, and I rather like how `flag.NewFlagSet(... ErrorHandling)` is set up.

How do you like this now?

**Comment by pellared** [comment]
@dertseha I will take a look later.
In the meantime can you change the branch name on your fork to `registered-config`?
Unfortunately, Go Modules are not happy with slashes in the branch name,
I wanted to play on your branch from a user perspective by adding:

```
replace github.com/pellared/taskflow => github.com/dertseha/taskflow feature/registered-config
```

to the `go.mod` file, but I get the following error:

```
replace github.com/dertseha/taskflow: version "feature/registered-config" invalid: version "feature/registered-config" invalid: disallowed version string
```

It would also let others (e.g. @sheldonhull) experiment and give feedback.

Reference: https://github.com/golang/go/wiki/Modules#when-should-i-use-the-replace-directive

**Comment by dertseha** [comment]
Welp, I did the renaming, apparently this also closes the PR. Continuing in #43 .
